### PR TITLE
Change how prefix and variable tags interact with each other

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Change the behavior of  `db::add_prefix<Prefix, Tag, Args...>` when `Tag` is a `::Tags::Variables<TagsList>` (with `TagsList = tmpl::list<Tag0,...,TagN>`) from producing a new tag `Prefix<::Tags::Variables<PrefixedTagsList, Args...>` with `PrefixedTagsList` = `tmpl::list<Prefix<Tag0, Args...>, ...,Prefix<TagN, Args...>>` to producing a new tag `::Tags::Variables<PrefixedTagsList, Args...>`.

I am making this change because it simplifies the code in `DataStructures/DataBox/PrefixHelpers.hpp` and because I found the tag `deriv<Variables<deriv<Var0>,...,deriv<VarN>>` to be potentially confusing as I read it to hold second derivatives...   In addition, there are several prefix tags that set `type = typename Tag::type` which is incorrect if `Tag` is a `::Tags::Variables`.

In cases where `db::add_prefix<Prefix, Tag, Args...>`  was never called on a `::Tags::Variables`, change to `Prefix<Tag, Args...>`.

In cases where `db::add_prefix<Prefix, Tag, Args...>` was always called on a
`::Tags::Variables`, rename the type alias to`db::variables_tag_with_tags_prefixed_by<Prefix, Tag, Args...>`

Remove `db::get_variables_tags_list` and `db::variables_tag_with_tags_list`
as they no longer need to handle a `::Tag::Variables` with both an outer and inner prefixes.

In addition, there are a few cleaning commits, and a couple of added `static_assert`s based on errors triggered when making these changes.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

@wthrowe there are two remaining uses of `db::add_tag_prefix` in `src/Time/Actions/RecordTimeStepperData.hpp` and `tests/Unit/Time/Actions/Test_SelfStartActions.cpp` because tests don't use a `Variables` as the `type` of the tag that is being wrapped.  Is this intentional, or can I update the tests to use a `Variables`?

@nilsleiffischer there are still many remaining uses of `db::add_tag_prefix` in the linear solver code, again because tests don't use a `Variables` as the `type` of the tag that is being wrapped.  Is this intentional?

### CODE REVIEW

**Comment by kidder** [comment]
- clang-tidy errors are from the boost root-finder that we have not been able to ignore
- mac build is broken (see #2397)

File: src/DataStructures/Variables.hpp
Comment (wthrowe) [inline src/DataStructures/Variables.hpp]: `flat_all_v`

File: src/Evolution/Actions/ComputeTimeDerivative.hpp
Comment (wthrowe) [inline src/Evolution/Actions/ComputeTimeDerivative.hpp]: This comment is out-of-date.  The action doesn't do any prefixing of anything anymore.

File: src/Time/Tags.hpp
Comment (wthrowe) [inline src/Time/Tags.hpp]: You don't use this.

**Review by wthrowe — CHANGES_REQUESTED** [review]
The use of the time-stepping code on non-Variables is intentional and I would prefer if it continued to work.

**Comment by wthrowe** [comment]
Thinking about this more, what I really want is to ensure that we can integrate a stack-based data structure in the future.  We've sort of been using double and array for that and I was thinking of adding a new container type with tags, but maybe it should be a "Variables\<double>" specialization.  If that idea seems sane, I retract my objection to the time changes.

File: src/ParallelAlgorithms/LinearSolver/Tags.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/LinearSolver/Tags.hpp:223]: Could you keep this comment around? E.g. add this:

```cpp
// No "Linear" prefix is added to the tag's `name()` because a Krylov subspace always refers to a linear operator
```

File: src/DataStructures/Tags/TempTensor.hpp
Comment (nilsvu) [inline src/DataStructures/Tags/TempTensor.hpp:1]: We don't have separate `Tags/` directories anywhere else in the code as far as I'm aware. What do you think about naming this file `DataStructures/TempTensorTags.hpp` instead?

File: tests/Unit/Helpers/DataStructures/TestTags.hpp
Comment (nilsvu) [inline tests/Unit/Helpers/DataStructures/TestTags.hpp]: Template this class on `UpLo` and `Dim`?

File: tests/Unit/Helpers/DataStructures/TestTags.hpp
Comment (nilsvu) [inline tests/Unit/Helpers/DataStructures/TestTags.hpp]: "Contract" implies another index to contract with, but the `type` just removes the first index. Name `RemoveFirstIndex` instead?

File: src/DataStructures/Variables.hpp
Comment (nilsvu) [inline src/DataStructures/Variables.hpp]: Can you replace `&&` -> `and` here and below?

File: src/DataStructures/DataBox/PrefixHelpers.hpp
Comment (nilsvu) [inline src/DataStructures/DataBox/PrefixHelpers.hpp]: Isn't this metafunction just `db::add_tag_prefix` with the restriction that it accepts nothing but a Variables tag? I think you can just delete it and we use `db::add_tag_prefix` everywhere.

File: src/Elliptic/Actions/InitializeSystem.hpp
Comment (nilsvu) [inline src/Elliptic/Actions/InitializeSystem.hpp]: You can leave these as-is, just using `db::add_tag_prefix` (same applies to many occurrences below)

(If your reasoning for this change is that you want to assert the tag is actually a variables tag: I think it's not the responsibility of the prefix utility metafunction to assert that. The class working with the tag, in this case `InitializeSystem`, should assert that what it receives is a variables tag if it expects one.)

File: src/NumericalAlgorithms/RootFinding/TOMS748.hpp
Comment (nilsvu) [inline src/NumericalAlgorithms/RootFinding/TOMS748.hpp]: This change seems unrelated to the rest of the commit. Better factor it into a separate commit if that's not too painful to do.

File: src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/LinearSolver/Gmres/Gmres.hpp]: nitpick: delete space, same in line 214

**Review by nilsvu — CHANGES_REQUESTED** [review]
The linear solver uses `db::add_tag_prefix` because it doesn't care if the tag holds a Variables or some vector type such as `DataVector` or `DenseVector`. It only requires the type supports addition, scalar multiplication and an inner product.

**Comment by nilsdeppe** [comment]
The idea of a `Variables<double>` specialization is something that I would be happy to see. I've thought a bit about that in the past, since we could continue to get vector instructions through Blaze that way by using the static vector class. Adding a `Requires` specilization on Variables for the tag type is probably the way to go unless someone has a better idea. That way at least the memory layout is hidden from users.

**Comment by wthrowe** [comment]
After looking into it, it looks like "Variables\<double>" (whether a specialization of Variables or not) is impossible to write without absurd tradeoffs.  We can't do Blaze math on Tensors because accessing the contents of multiple tensors as a big array violates C++'s stupid aliasing rules.  Allowing non-owning Tensors would work, but that has a memory overhead of more than a factor of 2.

I need to think a bit about how this affects what changes I'm OK with to the time stepping code.  It's nice to be able to integrate a few ODEs without having to go through the DataVector infrastructure (I've done it with our std::array math), but this now looks like it will be useless for any real runs.

**Comment by wthrowe** [comment]
Actually, I have one more idea.  I need to flesh out the details.

File: src/DataStructures/Tags/TempTensor.hpp
Comment (kidder) [reply src/DataStructures/Tags/TempTensor.hpp:1]: The plan is to move towards having separate `Tags/` directories with separate files per tag in order to reduce dependencies.  This is the first step...

File: tests/Unit/Helpers/DataStructures/TestTags.hpp
Comment (kidder) [reply tests/Unit/Helpers/DataStructures/TestTags.hpp]: I just removed this and `ContractFirst` as I don't use them

**Comment by kidder** [comment]
@wthrowe @nilsleiffischer I've pushed fixups.  In the interest of getting this merged sooner (as DataBox simplifications depend upon this) I've reverted the places I had replaced `add_tag_prefix` with `variables_tag_with_tags_prefixed_by` so they will work as before on either Variables or a non-Variables.  We can revisit this in the future as this does not affect the DataBox simplifications I am working on

File: src/Elliptic/Actions/InitializeSystem.hpp
Comment (nilsvu) [inline src/Elliptic/Actions/InitializeSystem.hpp]: The revert commit left over a bunch of these formatting changes

**Review by nilsvu — DISMISSED** [review]
Ok LGTM @kidder. Please don't wait for my re-approval if anything gets changed, since I'll be on vacation.

**Comment by wthrowe** [comment]
Looks good.  Squash, etc.

**Comment by kidder** [comment]
@wthrowe I've squashed, etc.

**Review by wthrowe — DISMISSED** [review]
Clang-tidy failure is Boost internals.

File: src/DataStructures/Tags.hpp
Comment (nilsdeppe) [inline src/DataStructures/Tags.hpp:88]: Need an include for tagname

**Review by nilsdeppe — COMMENTED** [review]
I'm happy with this. Just one missing include, the rest looks fine to me

**Comment by kidder** [comment]
@nilsdeppe @wthrowe fixed the include

**Review by nilsdeppe — DISMISSED** [review]

**Review by wthrowe — DISMISSED** [review]

**Review by nilsdeppe — APPROVED** [review]
CI error is clang-tidy in boost internals

**Review by wthrowe — APPROVED** [review]

**Pull request merged**
